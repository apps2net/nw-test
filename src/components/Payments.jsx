import React from "react"
import PropTypes from 'prop-types';
import { Table, ButtonGroup, Button, ToggleButton } from "react-bootstrap";

const Payments = (props) => {

    const { payments, isLoading, paymentStatus, controlFunc, statusControlFunc } = props;

    const radios = [
        { value: '1', name: 'All' },
        { value: 'A', name: 'Approved' },
        { value: 'C', name: 'Cancelled' },
        { value: 'P', name: 'Pending' },
    ];

    /* Converts above array to an object for mapping status code to display string */
    function objectify(array) {
        return array.reduce(function(result, currentArray) {
            result[currentArray.value] = currentArray.name;
            return result;
        }, {});
    }
    
    const paymentStatusObj = (objectify(radios));
    
    /* Define a style for the heading - just an alternative to using app.css*/
    const h1Style = {
        color: 'blue',
        fontSize: '2rem'
        
      };

    return (
        <React.Fragment>
            <div role="main">
            <h1 style={h1Style}>Payment Listing</h1>
            <div>
                <ButtonGroup className="mb-2">
                    {radios.map((radio, idx) => (
                        <ToggleButton
                            key={idx}
                            id={`radio-${idx}`}
                            type="radio"
                            variant="secondary"
                            name="radio"
                            value={radio.value}
                            checked={paymentStatus === radio.value}
                            onChange={(e) => statusControlFunc(e.currentTarget.value)}
                        >
                            {radio.name}
                        </ToggleButton>
                    ))}
                </ButtonGroup>
            </div>
            <div>
                <Table id="payments-table" data-testid="payments-table" striped bordered hover size="sm">
                    <thead>
                        <tr>
                            <th>Amount</th>
                            <th>Currency</th>
                            <th>Date</th>
                            <th>Status</th>
                            <th>Type</th>
                            <th>From</th>
                            <th>To</th>
                        </tr>
                    </thead>
                    <tbody>
                        {payments.results && (
                            payments.results.map((payment, index) => {
                                if (paymentStatus === '1' || payment.paymentStatus === paymentStatus) {
                                    return (
                                        <tr key={"payment-" + index} data-testid="payments-row">
                                            <td>{payment.paymentAmount}</td>
                                            <td>{payment.paymentCurrency}</td>
                                            <td>{payment.paymentDate}</td>
                                            <td>{paymentStatusObj[payment.paymentStatus]}</td>
                                            <td>{payment.paymentType}</td>
                                            <td>{payment.fromAccount.accountName}<br />{payment.fromAccount.accountNumber} - {payment.fromAccount.sortCode}</td>
                                            <td>{payment.toAccount.accountName}<br />{payment.toAccount.accountNumber} - {payment.toAccount.sortCode}</td>
                                        </tr>
                                    )
                                }
                            }
                            )
                        )}
                    </tbody>
                </Table>
                <div>
                <Button
                    variant="primary"
                    disabled={isLoading||!payments.metaData.hasMoreElements}
                    onClick={!isLoading ? () => controlFunc(payments.metaData.nextPageIndex) : null}
                >
                    {isLoading ? 'Loadingâ€¦' : 'Load more'}
                </Button>
                </div>
            </div>
            </div>
        </React.Fragment>

    );
}

Payments.propTypes = {
    payments: PropTypes.object.isRequired,
    isLoading: PropTypes.bool,
    paymentStatus: PropTypes.string.isRequired,
    controlFunc: PropTypes.func.isRequired,
    statusControlFunc: PropTypes.func.isRequired
};

Payments.defaultProps = {
    isLoading: false,
    paymentStatus: '1'
}

export default Payments