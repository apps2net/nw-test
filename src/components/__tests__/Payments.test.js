
import React from "react";
import {document, render, screen, cleanup} from '@testing-library/react';
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom'
import Payments from '../Payments';

/**
 * @jest-environment jsdom
 */


describe('Payments Table Test', () => {

	const paymentData = {
		"results": [{ "paymentAmount": "558.52", "paymentCurrency": "GBP", "paymentType": "FASTER", "paymentDate": "24-Sep-2021", "paymentStatus": "P", 
		"toAccount": { "accountName": "Investment Account", "sortCode": "381692", "accountNumber": "63385533" }, 
		"fromAccount": { "accountName": "Checking Account", "sortCode": "521698", "accountNumber": "09189790" } }, 
		{ "paymentAmount": "157.79", "paymentCurrency": "GBP", "paymentType": "FASTER", "paymentDate": "24-Sep-2021", "paymentStatus": "C", 
		"toAccount": { "accountName": "Investment Account", "sortCode": "374903", "accountNumber": "55339636" }, "fromAccount": { "accountName": "Credit Card Account", "sortCode": "775809", "accountNumber": "19862324" } }
		], "metaData": { "hasMoreElements": true, "nextPageIndex": "qJGSEcKL" }
	};

	const controlFunc = (val) => {
        console.log(val)
    }


	// Applies only to tests in this describe block
	beforeEach(() => {
		return;
	});

	afterEach(() => {
		cleanup();
	})

	test(' has a valid snapshot', () => { 
		const component = renderer.create( <Payments
            payments={paymentData}
            isLoading={false}
            paymentStatus={'1'}
            controlFunc={controlFunc}
            statusControlFunc={controlFunc} /> ); 
		let tree = component.toJSON(); 
		expect( tree).toMatchSnapshot(); 
	});


	test('should render Payments component', () => {
		render(
			<Payments
            payments={paymentData}
            isLoading={false}
            paymentStatus={'1'}
            controlFunc={controlFunc}
            statusControlFunc={controlFunc} />				
			)

			/*  Check the load more button is there */
			expect(screen.getByText('Load more')).toBeInTheDocument();
			/* Check the account number is listed */
			expect(screen.getByText(/55339636/)).toBeInTheDocument();
			/* Check for two rows plus the header */
			expect(screen.getAllByRole('row')).toHaveLength(3)
	});

});