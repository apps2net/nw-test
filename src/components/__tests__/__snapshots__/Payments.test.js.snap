// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payments Table Test  has a valid snapshot 1`] = `
<div
  role="main"
>
  <h1
    style={
      Object {
        "color": "blue",
        "fontSize": "2rem",
      }
    }
  >
    Payment Listing
  </h1>
  <div>
    <div
      className="mb-2 btn-group"
      role="group"
    >
      <label
        className="btn active btn-secondary"
        disabled={false}
        id="radio-0"
      >
        <input
          autoComplete="off"
          checked={true}
          disabled={false}
          name="radio"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="radio"
          value="1"
        />
        All
      </label>
      <label
        className="btn btn-secondary"
        disabled={false}
        id="radio-1"
      >
        <input
          autoComplete="off"
          checked={false}
          disabled={false}
          name="radio"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="radio"
          value="A"
        />
        Approved
      </label>
      <label
        className="btn btn-secondary"
        disabled={false}
        id="radio-2"
      >
        <input
          autoComplete="off"
          checked={false}
          disabled={false}
          name="radio"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="radio"
          value="C"
        />
        Cancelled
      </label>
      <label
        className="btn btn-secondary"
        disabled={false}
        id="radio-3"
      >
        <input
          autoComplete="off"
          checked={false}
          disabled={false}
          name="radio"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="radio"
          value="P"
        />
        Pending
      </label>
    </div>
  </div>
  <div>
    <table
      className="table table-sm table-striped table-bordered table-hover"
      data-testid="payments-table"
      id="payments-table"
    >
      <thead>
        <tr>
          <th>
            Amount
          </th>
          <th>
            Currency
          </th>
          <th>
            Date
          </th>
          <th>
            Status
          </th>
          <th>
            Type
          </th>
          <th>
            From
          </th>
          <th>
            To
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="payments-row"
        >
          <td>
            558.52
          </td>
          <td>
            GBP
          </td>
          <td>
            24-Sep-2021
          </td>
          <td>
            Pending
          </td>
          <td>
            FASTER
          </td>
          <td>
            Checking Account
            <br />
            09189790
             - 
            521698
          </td>
          <td>
            Investment Account
            <br />
            63385533
             - 
            381692
          </td>
        </tr>
        <tr
          data-testid="payments-row"
        >
          <td>
            157.79
          </td>
          <td>
            GBP
          </td>
          <td>
            24-Sep-2021
          </td>
          <td>
            Cancelled
          </td>
          <td>
            FASTER
          </td>
          <td>
            Credit Card Account
            <br />
            19862324
             - 
            775809
          </td>
          <td>
            Investment Account
            <br />
            55339636
             - 
            374903
          </td>
        </tr>
      </tbody>
    </table>
    <div>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Load more
      </button>
    </div>
  </div>
</div>
`;
